lib/rules/id-match.js

```diff
@@ -106,36 +143,70 @@ module.exports = {
+                    // Report AssignmentExpressions left side's assigned variable id
                     } else if (effectiveParent.type === "AssignmentExpression" &&
-                        (effectiveParent.right.type !== "MemberExpression" ||
                         effectiveParent.left.type === "MemberExpression" &&
-                        effectiveParent.left.property.name === name)) {
+                        effectiveParent.left.property.name === node.name) {
+                        if (isInvalid(name)) {
+                            report(node);
+                        }
+
+                    // Report AssignmentExpressions only if they are the left side of the assignment
+                    } else if (effectiveParent.type === "AssignmentExpression" && effectiveParent.right.type !== "MemberExpression") {
                         if (isInvalid(name)) {
                             report(node);
                         }
                     }
 
-                } else if (parent.type === "Property") {
+                /*
+                 * Properties have their own rules, and
+                 * AssignmentPattern nodes can be treated like Properties:
+                 * e.g.: const { no_camelcased = false } = bar;
+                 */
+                } else if (parent.type === "Property" || parent.type === "AssignmentPattern") {
+
+                    if (parent.parent && parent.parent.type === "ObjectPattern") {
+                        if (parent.shorthand && parent.value.left && isInvalid(name)) {
+
+                            report(node);
+                        }
+
+                        const assignmentKeyEqualsValue = parent.key.name === parent.value.name;
 
-                    if (!properties || parent.key.name !== name) {
+                        // prevent checking righthand side of destructured object
+                        if (!assignmentKeyEqualsValue && parent.key === node) {
+                            return;
+                        }
+
+                        const valueIsInvalid = parent.value.name && isInvalid(name);
+
+                        // ignore destructuring if the option is set, unless a new identifier is created
+                        if (valueIsInvalid && !(assignmentKeyEqualsValue && ignoreDestructuring)) {
+                            report(node);
+                        }
+                    }
+
+                    // never check properties or always ignore destructuring
+                    if (!properties || (ignoreDestructuring && isInsideObjectPattern(node))) {
                         return;
                     }
 
-                    if (shouldReport(effectiveParent, name)) {
+                    // don't check right hand side of AssignmentExpression to prevent duplicate warnings
+                    if (parent.right !== node && shouldReport(effectiveParent, name, true)) {
```

Comment:
seems the 3rd  param `true` is not unused?

Reply:
My mistake, this param is not used for this commit.